evm/contracts/examples/ConversionRate.sol

```diff
@@ -107,65 +132,142 @@ contract ConversionRate is ChainlinkClient, Ownable {
   }
 
   /**
    * @dev Performs aggregation of the answers received from the Chainlink nodes.
+   * Assumes that at least half the oracles are honest and so can't contol the
+   * middle of the ordered responses.
    * @param _answerId The answer ID associated with the group of requests
    */
-  function updateRate(uint256 _answerId)
+  function updateLatestAnswer(uint256 _answerId)
     private
+    ensureMinResponsesReceived(_answerId)
     ensureOnlyLatestAnswer(_answerId)
   {
-    uint256 sumQuotients;
-    uint256 sumRemainders;
     Answer memory answer = answers[_answerId];
-    for (uint i = 0; i < answer.expectedResponses; i++) {
-      uint256 response = answer.responses[i];
-      sumQuotients = sumQuotients.add(response.div(answer.expectedResponses)); // aggregate responses and protect from overflows
-      sumRemainders = sumRemainders.add(response % answer.expectedResponses);
+    uint256 responseLength = answer.responses.length;
+    uint256 middleIndex = responseLength.div(2);
+    if (responseLength % 2 == 0) {
+      uint256 median1 = quickselect(answers[_answerId].responses, middleIndex);
+      uint256 median2 = quickselect(answers[_answerId].responses, middleIndex.add(1)); // quickselect is 1 indexed
+      currentRate = median1.add(median2).div(2);
+    } else {
+      currentRate = quickselect(answers[_answerId].responses, middleIndex.add(1)); // quickselect is 1 indexed
     }
-    currentRate = sumQuotients.add(sumRemainders.div(answer.expectedResponses)); // recover lost accuracy from result 
     latestCompletedAnswer = _answerId;
   }
 
   /**
-   * @notice Called by the owner to kill the contract. This transfers all LINK
-   * balance and ETH balance (if there is any) to the owner.
+   * @dev Returns the kth value of the ordered array
+   * See: http://www.cs.yale.edu/homes/aspnes/pinewiki/QuickSelect.html
+   * @param _a The list of elements to pull from
+   * @param _k The index, 1 based, of the elements you want to pull from when ordered
    */
-  function destroy()
-    public
-    onlyOwner
+  // solium-disable-next-line security/no-assign-params
+  function quickselect(uint256[] memory _a, uint256 _k)
+    private
+    pure
+    returns (uint256)
   {
-    LinkTokenInterface link = LinkTokenInterface(chainlinkTokenAddress());
-    transferLINK(owner, link.balanceOf(address(this)));
-    selfdestruct(owner);
+    uint256 aLen = _a.length;
+    uint256[] memory a1 = new uint256[](aLen);
+    uint256[] memory a2 = new uint256[](aLen);
+    uint256 a1Len;
+    uint256 a2Len;
+    uint256 pivot;
+    uint256 i;
+
+    while (true) {
+      pivot = _a[aLen.div(2)];
+      a1Len = 0;
+      a2Len = 0;
+      for (i = 0; i < aLen; i++) {
+        if (_a[i] < pivot) {
+          a1[a1Len] = _a[i];
+          a1Len++;
+        } else if (_a[i] > pivot) {
+          a2[a2Len] = _a[i];
+          a2Len++;
+        }
+      }
+      if (_k <= a1Len) {
+        aLen = a1Len;
+        for (i = 0; i < a1Len; i++) {
```

Comment:
Can probably save a bit of gas and LOC, here and in the corresponding code for `k > blah` below, by just saying `_a = a1`.

Reply:
Great suggestion.